ddht/v5_1/client.py

```diff
@@ -460,3 +486,93 @@ async def topic_query(
         request_id: Optional[bytes] = None,
     ) -> InboundMessage[FoundNodesMessage]:
         raise NotImplementedError
+
+
+@asynccontextmanager
+async def common_client_stream_find_nodes(
+    client: Client,
+    node_id: NodeID,
+    endpoint: Endpoint,
+    distances: Collection[int],
+    *,
+    request_id: Optional[bytes] = None,
+) -> AsyncIterator[trio.abc.ReceiveChannel[InboundMessage[FoundNodesMessage]]]:
+    async def _stream_find_nodes_response(
+        response_message_type: Type[BaseMessage],
+        send_channel: trio.abc.SendChannel[InboundMessage[FoundNodesMessage]],
+    ) -> None:
+
+        head_response_total = 0
+        response_counter = 0
+
+        with trio.move_on_after(REQUEST_RESPONSE_TIMEOUT) as scope:
+            async with send_channel:
+                with client.request_tracker.reserve_request_id(
+                    node_id, request_id
+                ) as reserved_request_id:
+                    request = AnyOutboundMessage(
+                        FindNodeMessage(reserved_request_id, tuple(distances)),
+                        endpoint,
+                        node_id,
+                    )
+
+                    async with client.dispatcher.subscribe_request(
+                        request, response_message_type
+                    ) as subscription:
+                        async for response in subscription:
+                            if response.message.total == 0:
+                                raise ValidationError(
+                                    "Invalid `total` counter in response: "
+                                    f"total={response.message.total}"
+                                )
+
+                            if response_counter == 0:
+                                head_response_total = response.message.total
```

Comment:
I'm curious what you think about the following pattern.

```python
head_response = await subscription.receive()
yield head_response

total_expected_messages = head_response.message.total

for _ in range(total_expected_messages - 1):  # since we already received one...
    response = await subscription.receive()
    yield response
```

I like that it gets rid of this logic for detecting that it is our first pass through the loop.  I think it trades that for a little bit of awkwardness of not actually using the channel as an async iterator which is a very natural way to do things and requires instead that we explicitely call `receive()`.  Thoughts?  I'm not convinced the structure above is better.

Reply:
It's not obvious to me that one option's better than the other. But I prefer removing the unnecessarily run logic over whatever slight gain in readability from using the channel "properly" (since imo it's still very obvious as to what's happening)

Reply:
Now that I'm tinkering with it - we're not just `yield response`ing here, but sending the response over the `send_channel`, which comes with it's own `try/except` and doing some validation, so in this case it seems a little to cumbersome duplicating all that logic to make whatever slight gain in performance worth it.

Reply:
Here's my implementation of roughly the same logic for the Alexandria APIs that have roughly the same shape:

https://github.com/ethereum/ddht/pull/278/files

I didn't end up putting any `try/except` around the channel operations.  I'm not 100% sure they aren't needed, but I *think* that it might not be necessary.  I'd at least look into removing the handling and see if you get errors in a few test runs locally and in CI.  If nothing blows up :+1:  and if it blows up later we know where to add the handling.